### PR TITLE
Gh 191/Transactions list networking

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/Logic/Models/Transaction.swift [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+//  Multisig
+//
+//  Created by Moaaz on 6/1/20.
+//  Copyright © 2020 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import BigInt
+
+
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Transaction.swift:12]: blank spaces

File: Multisig/UI/Assets/BalancesViewModel.swift [hunk 2]
```diff
@@ -48,6 +44,18 @@ struct TokenBalance: Identifiable, Hashable {
...
 }
 
+extension TokenFormatter {
+
+    func safeString(from: String?, decimals: Int? = nil, isNegative: Bool = false) -> String {
+        string(
+            from: BigDecimal((isNegative ? -1 : 1) * (BigInt(from ?? "0") ?? 0), decimals ?? 18),
+            decimalSeparator: Locale.autoupdatingCurrent.decimalSeparator ?? ".",
+            thousandSeparator: Locale.autoupdatingCurrent.groupingSeparator ?? ",",
+            forcePlusSign: true
```
Comment (DmitryBespalov) [inline Multisig/UI/Assets/BalancesViewModel.swift]: please make this as another parameter to this function, with default value = true, but call it from the line 33 above with value of false

File: Multisig/UI/Transactions/TransactionsViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    var dataLength: Int
+
+    override init() {
+        dataLength  = 98
+        super.init()
+    }
+}
+
+
+//func cellForModel(_ m: BaseTransactionViewModel) {
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionsViewModel.swift]: this comment needs to be removed

File: Multisig/Cross-layer/App.swift [hunk 4]
```diff
@@ -25,13 +25,17 @@ class App {
     // Data Layer
 
     // Services    
-    var safeTransactionService = SafeTransactionService(url: URL(string: "https://safe-transaction.rinkeby.gnosis.io")!,
+    var safeTransactionService = SafeTransactionService(url: URL(string: "https://safe-transaction.staging.gnosisdev.com/")!,
     logger: LogService.shared)
 
+    var safeRelayService = SafeRelayService(url: URL(string: "https://safe-relay.staging.gnosisdev.com")!, logger: LogService.shared)
```
Comment (sche) [inline Multisig/Cross-layer/App.swift:31]: Did we add to the list of services in the app settings advanced section?

File: Multisig/Data/Services/SafeRelayService.swift [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            symbol: symbol ?? "",
+            description: nil,
+            decimals: decimals ?? 0,
+            websiteUri: nil,
+            gas: nil)
+
+    }
+
+    subscript(address: String) -> TokensRequest.Token? {
+        guard let address = Address(address) else { return nil }
```
Comment (sche) [inline Multisig/Data/Services/SafeRelayService.swift:95]: So we assume that this subscript should always be requested from a background queue, right?

Comment (MouazAlzahabi) [reply Multisig/Data/Services/SafeRelayService.swift:95]: now I did

File: Multisig/Data/Services/SafeTransactionService.swift [hunk 6]
```diff
@@ -78,3 +105,28 @@ struct SafeBalancesRequest: JSONRequest {
...
+
+    init(address: String) {
+        self.address = address
+    }
+
+    var query: String? {
+        return "limit=\(limit)&offset=\(offset)"
+    }
+
+    struct Response: Decodable {
```
Comment (sche) [inline Multisig/Data/Services/SafeTransactionService.swift:126]: can we reuse 'PagedResponse' from the SafeRelayService file?

File: Multisig/Cross-layer/App.swift (hunk 4 above)
Comment (MouazAlzahabi) [reply Multisig/Cross-layer/App.swift:31]: now I did

Comment (DmitryBespalov) [reply Multisig/Cross-layer/App.swift:31]: no, we didn't, this is a temporary solution, though

File: Multisig/Data/Services/SafeRelayService.swift (hunk 5 above)
Comment (DmitryBespalov) [reply Multisig/Data/Services/SafeRelayService.swift:95]: yes

File: Multisig/Data/Services/SafeTransactionService.swift (hunk 6 above)
Comment (DmitryBespalov) [reply Multisig/Data/Services/SafeTransactionService.swift]: yes, the `PagedResponse` should be used here, too

File: Multisig/Logic/Models/Transaction.swift [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+extension Transaction {
+
+    func status(safeNonce: Int, safeThreshold: Int) -> TransactionStatus {
+        // tx-es without nonce are external transactions that are
+        // already executed successfully.
+        guard let nonce = nonce else { return .success }
+        let confirmationCount = confirmations?.count ?? 0
+        let threshold = confirmationsRequired ?? safeThreshold
```
Comment (sche) [inline Multisig/Logic/Models/Transaction.swift:58]: Can confirmationsRequired be ever optional?

File: Multisig/Logic/Models/Transaction.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+        if isExecuted == true && isSuccessful == true {
+            return .success
+        } else if isExecuted == true && isSuccessful != true {
```
Comment (sche) [inline Multisig/Logic/Models/Transaction.swift:62]: do we have unit tests for all these cases?

Comment (DmitryBespalov) [reply Multisig/Logic/Models/Transaction.swift:62]: no, we don't, nothing is unit tested

File: Multisig/Logic/Models/Transaction.swift (hunk 7 above)
Comment (MouazAlzahabi) [reply Multisig/Logic/Models/Transaction.swift:58]: Yes I faced this case

Comment (DmitryBespalov) [reply Multisig/Logic/Models/Transaction.swift:58]: yes, for some type of transactions this field is not existing (`txType == ETHEREUM_TRANSACTION`)

File: Multisig/Logic/Models/Transaction.swift [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    let parameters: [DecodedDataParameter]
+}
+
+struct DecodedDataParameter: Decodable, Hashable {
+    let name: String
+    let type: String
+    let value: String
+}
+
+enum TransactionType: String, Decodable {
```
Comment (sche) [inline Multisig/Logic/Models/Transaction.swift:109]: can we add somewhere a link to the doc describing response?

File: Multisig/UI/Transactions/TransactionCell/TransactionCellView.swift [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        let transferTransaction = transaction as? TransferTransaction
+        let settingChangeTransaction = transaction as? SettingChangeTransaction
+        let changeMasterCopyTransaction = transaction as? ChangeMasterCopyTransaction
+        let customTransaction = transaction as? CustomTransaction
+
+        return ZStack {
+            if customTransaction != nil {
+                CustomTransactionCellView(transaction: customTransaction!)
+            }
+            else if transferTransaction != nil {
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionCell/TransactionCellView.swift:53]: should we put "else if" after "}" ?

File: Multisig/UI/Transactions/TransactionsViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+    }
+}
+
+
+class TransactionsViewModel: ObservableObject {
+    @Published var transactionsList = TransactionsList()
+
+    @Published var isLoading: Bool = true
+    @Published var errorMessage: String? = nil
+
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionsViewModel.swift:116]: extra space

File: Multisig/UI/Transactions/TransactionsViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+    func loadData() {
+        isLoading = true
+        Just(safe.address!)
+            .setFailureType(to: Error.self)
+            .flatMap { address in
+                Future<TransactionsList, Error> { promise in
+                    DispatchQueue.global().async {
+                        do {
+                            let info = try App.shared.safeTransactionService.safeInfo(at: address)
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionsViewModel.swift:134]: Is it a temporary solution? I thought that we should have all the required info from the endpoint.

File: Multisig/UI/Transactions/TransactionsViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+        Self.queueStatuses.contains(self)
+    }
+
+    var isInHistory: Bool {
+        Self.historyStatuses.contains(self)
+    }
+}
+
+
+class TransactionsViewModel: ObservableObject {
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionsViewModel.swift:111]: we should create a tech debt task to create unit tests for this view model. And for other mentioned places too.

File: Multisig/UI/Transactions/TransactionsViewModel.swift (hunk 12 above)
Comment (MouazAlzahabi) [reply Multisig/UI/Transactions/TransactionsViewModel.swift:134]: There is open ticket for that, but it seems we have to live with this way for the time being

**Pull request merged**